### PR TITLE
Send event on referral engine state change

### LINKED ISSUE #9046: Send event on referral engine state change

### PR DESCRIPTION
**Pull request by ValentinTrinque**
Close #9046

### CODE REVIEW

File: protos/sources/vega/events/v1/events.proto [hunk 1]
```diff
@@ -727,6 +727,23 @@ message RefereeJoinedTeam {
   int64 joined_at = 3;
 }
 
+message ReferralProgramStarted {
+  // Program that started.
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:731]:
```suggestion
  // Referral program that started.
```

File: protos/sources/vega/events/v1/events.proto [hunk 2]
```diff
@@ -727,6 +727,23 @@ message RefereeJoinedTeam {
...
+  vega.ReferralProgram program = 1;
+}
+
+message ReferralProgramUpdated {
+  // The updated program.
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:736]:
```suggestion
  // The updated referral program.
```

File: protos/sources/vega/events/v1/events.proto [hunk 3]
```diff
@@ -727,6 +727,23 @@ message RefereeJoinedTeam {
...
+  vega.ReferralProgram program = 1;
+}
+
+message ReferralProgramEnded {
+  // Program update version.
+  uint64 version = 1;
+  // Unique identifier of the program.
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:743]:
```suggestion
  // Unique ID of the referral program.
```

File: protos/sources/vega/events/v1/events.proto [hunk 4]
```diff
@@ -1018,6 +1045,12 @@ message BusEvent {
     RefereeSwitchedTeam referee_switched_team = 163;
     // Event notifying that a referee joined a team.
     RefereeJoinedTeam referee_joined_team = 164;
+    // Event notifying a referral program started.
+    ReferralProgramStarted referral_program_started = 165;
+    // Event notifying a referral program has been updated.
+    ReferralProgramUpdated referral_program_updated = 166;
+    // Event notifying a referral program ended.
+    ReferralProgramEnded referral_program_ended = 167;
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:1053]:
```suggestion
    // Event notifying that a referral program has started.
    ReferralProgramStarted referral_program_started = 165;
    // Event notifying that a referral program has been updated.
    ReferralProgramUpdated referral_program_updated = 166;
    // Event notifying that a referral program has ended.
    ReferralProgramEnded referral_program_ended = 167;
```

File: protos/sources/vega/vega.proto [hunk 5]
```diff
@@ -1573,14 +1573,19 @@ message MatrixValue {
 }
 
 message ReferralProgram {
+  // Incremental version of the program. It is incremented after each program
+  // update.
+  uint64 version = 1;
+  // Unique identifier generated from the proposal that created this program.
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:1579]:
```suggestion
  // Unique ID generated from the proposal that created this program.
```

File: protos/sources/vega/events/v1/events.proto [hunk 6]
```diff
@@ -727,6 +727,23 @@ message RefereeJoinedTeam {
   int64 joined_at = 3;
 }
 
+message ReferralProgramStarted {
+  // Referral program that started.
```
Comment (gordsport) [inline protos/sources/vega/events/v1/events.proto:731]:
```suggestion
  // Referral program that has started.
```

Comment (gordsport) [reply protos/sources/vega/events/v1/events.proto]: or "The referral program that has started"

**Review by gordsport — DISMISSED** [review]

File: core/referral/engine.go [hunk 7]
```diff
@@ -1,24 +1,44 @@
...
+	// twice.
+	latestProgramVersion uint64
+
 	// currentProgram is the program currently in used against which the reward
 	// are computed.
 	// It's `nil` is there is none.
 	currentProgram *types.ReferralProgram
 
-	// currentProgramHasEnded tells is the current program has reached its
+	// programHasEnded tells is the current program has reached its
```
Comment (peterbarrow) [inline core/referral/engine.go:39]:
```suggestion
	// programHasEnded tells if the current program has reached it's
```

File: core/referral/engine.go [hunk 8]
```diff
@@ -27,55 +47,78 @@ type Engine struct {
...
+		}
 	}
 
-	// Verifying if the new program is not already finished.
-	if e.newProgram != nil && e.newProgram.EndOfProgramTimestamp.Before(epochEnd) {
-		e.newProgram = nil
-		// TODO: Send event to tell the new program didn't event start.
+	// This handles a edge case where the new program ends before the next
+	// epoch starts. It can happen when the proposal updating the referral
+	// program doesn't specify an end date that is to clause to the enactment
```
Comment (peterbarrow) [inline core/referral/engine.go:80]:
```suggestion
	// program doesn't specify an end date that is to close to the enactment
```

**Review by peterbarrow — DISMISSED** [review]

**Review by gordsport — DISMISSED** [review]

**Review by candida-d — DISMISSED** [review]

**Review by guoguojin — DISMISSED** [review]

**Review by candida-d — APPROVED** [review]

**Review by peterbarrow — APPROVED** [review]

**Pull request merged**